### PR TITLE
group_split(), group_nest(), group_keys(), group_map()

### PR DESCRIPTION
**Pull request by romainfrancois**
This is #3949 's sister pr, experimenting with using the `group_` prefix: 

This adds the three generics: 
 - `group_split()` : splits the tibble according to a grouping spec
 - `group_nest()` : nest a tibble by a grouping spec
 - `group_keys()` : grouping keys from a grouping spec

Each have methods for: 
 - data frame: in that case, the grouping spec comes from `...` and is forwarded to `group_by()`
 - grouped_df: the `...` is ignored, with a warning that says you should `group_by(add = TRUE)` first if this is really what you want to to

they also have their colwise variants: `group_split_at`, `group_split_if`, `group_nest_at`, `group_nest_if`, `group_keys_at`, `group_keys_if`. 

I think this is very natural, the story becomes that the `group_` family is extended. 

Not sure about the name `group_keys` because as opposed to the other two, `keys` is not a verb. 

``` r
library(dplyr)
# with an ungrouped data frame and a grouping spec
iris %>% group_nest(Species)
#> # A tibble: 3 x 2
#>   Species    data             
#>   <fct>      <list>           
#> 1 setosa     <tibble [50 × 4]>
#> 2 versicolor <tibble [50 × 4]>
#> 3 virginica  <tibble [50 × 4]>
iris %>% group_split(Species)
#> [[1]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 40 more rows
#> 
#> [[2]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>     
#>  1          7           3.2          4.7         1.4 versicolor
#>  2          6.4         3.2          4.5         1.5 versicolor
#>  3          6.9         3.1          4.9         1.5 versicolor
#>  4          5.5         2.3          4           1.3 versicolor
#>  5          6.5         2.8          4.6         1.5 versicolor
#>  6          5.7         2.8          4.5         1.3 versicolor
#>  7          6.3         3.3          4.7         1.6 versicolor
#>  8          4.9         2.4          3.3         1   versicolor
#>  9          6.6         2.9          4.6         1.3 versicolor
#> 10          5.2         2.7          3.9         1.4 versicolor
#> # … with 40 more rows
#> 
#> [[3]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>    
#>  1          6.3         3.3          6           2.5 virginica
#>  2          5.8         2.7          5.1         1.9 virginica
#>  3          7.1         3            5.9         2.1 virginica
#>  4          6.3         2.9          5.6         1.8 virginica
#>  5          6.5         3            5.8         2.2 virginica
#>  6          7.6         3            6.6         2.1 virginica
#>  7          4.9         2.5          4.5         1.7 virginica
#>  8          7.3         2.9          6.3         1.8 virginica
#>  9          6.7         2.5          5.8         1.8 virginica
#> 10          7.2         3.6          6.1         2.5 virginica
#> # … with 40 more rows
iris %>% group_keys(Species)
#> # A tibble: 3 x 1
#>   Species   
#>   <fct>     
#> 1 setosa    
#> 2 versicolor
#> 3 virginica

# with an already grouped df
iris %>% group_by(Species) %>% group_nest()
#> # A tibble: 3 x 2
#>   Species    data             
#>   <fct>      <list>           
#> 1 setosa     <tibble [50 × 4]>
#> 2 versicolor <tibble [50 × 4]>
#> 3 virginica  <tibble [50 × 4]>
iris %>% group_by(Species) %>% group_split()
#> [[1]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 40 more rows
#> 
#> [[2]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>     
#>  1          7           3.2          4.7         1.4 versicolor
#>  2          6.4         3.2          4.5         1.5 versicolor
#>  3          6.9         3.1          4.9         1.5 versicolor
#>  4          5.5         2.3          4           1.3 versicolor
#>  5          6.5         2.8          4.6         1.5 versicolor
#>  6          5.7         2.8          4.5         1.3 versicolor
#>  7          6.3         3.3          4.7         1.6 versicolor
#>  8          4.9         2.4          3.3         1   versicolor
#>  9          6.6         2.9          4.6         1.3 versicolor
#> 10          5.2         2.7          3.9         1.4 versicolor
#> # … with 40 more rows
#> 
#> [[3]]
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>    
#>  1          6.3         3.3          6           2.5 virginica
#>  2          5.8         2.7          5.1         1.9 virginica
#>  3          7.1         3            5.9         2.1 virginica
#>  4          6.3         2.9          5.6         1.8 virginica
#>  5          6.5         3            5.8         2.2 virginica
#>  6          7.6         3            6.6         2.1 virginica
#>  7          4.9         2.5          4.5         1.7 virginica
#>  8          7.3         2.9          6.3         1.8 virginica
#>  9          6.7         2.5          5.8         1.8 virginica
#> 10          7.2         3.6          6.1         2.5 virginica
#> # … with 40 more rows
iris %>% group_by(Species) %>% group_keys()
#> # A tibble: 3 x 1
#>   Species   
#>   <fct>     
#> 1 setosa    
#> 2 versicolor
#> 3 virginica
```

<sup>Created on 2018-11-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
Following this, we could have a `group_map` that would allow iteration over groups, e.g. 

```
iris %>%
  group_by(Species) %>%
  group_map( ~ lambda)
```

where lambda would be a function of: 
  - `.key` : a 1-row tibble identifying the group
  - `.subset` : the subset of `.data` for that group

File: R/group_keys.R [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+#' The grouping keys
+#'
+#' The grouping keys are a tibble with only the grouping information.
+#'
+#' @param .data A tbl
+#' @param ... See [group_by()], [group_by_at()] or [group_by_if()]
+#'
+#' @return A tibble with one row per group and one column per grouping variable
```
Comment (hadley) [inline R/group_keys.R]: These should probably have a `@family` so they all get cross-referenced

File: R/group_keys.R [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#'   group_keys()
+#'
+#' @export
+group_keys <- function(.data, ...) {
+  UseMethod("group_keys")
+}
+
+#' @export
+group_keys.default <- function(.data, ...){
+  abort("unsupported")
```
Comment (hadley) [inline R/group_keys.R]: If you're going to supply a default method, you need to provide an informative error message.

File: R/group_keys.R [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+#' @export
+group_keys.data.frame <- function(.data, ...){
+  group_keys_impl(group_by(.data, ...))
+}
+
+#' @export
+group_keys.grouped_df <- function(.data, ...) {
+  if (dots_n(...)) {
+    warn("... is not supported in group_keys(<grouped_df>), consider group_by(..., add = TRUE) %>% group_split()")
```
Comment (hadley) [inline R/group_keys.R]: is not supported -> ignored. consider -> please use

File: R/group_nest.R [hunk 4]
```diff
@@ -0,0 +1,96 @@
+
+group_nest_impl <- function(.data, key_var){
+  to_nest <- select(ungroup(.data), setdiff(tbl_vars(.data), group_vars(.data)))
+
+  out <- .data %>%
```
Comment (hadley) [inline R/group_nest.R]: I'd prefer to not use `%>%` just because it currently makes the backtraces harder to read

File: R/group_split.R [hunk 5]
```diff
@@ -0,0 +1,68 @@
+#' Split data frame by groups
+#'
```
Comment (hadley) [inline R/group_split.R]: I think this should the differences with `split()`, and mention `group_keys()` as a way to get the key information.

**Comment by hadley** [comment]
Also should `group_nest()` be implemented with `group_keys()` and `group_split()`?

File: R/group_nest.R (hunk 4 above)
Comment (romainfrancois) [reply R/group_nest.R]: Yeah, it's a proof of concept code that has been there since we first talked about nest_by.

File: R/group_keys.R (hunk 2 above)
Comment (romainfrancois) [reply R/group_keys.R]: For some reason, I though I had to have the `.default`. Removed the default for now.

File: R/group_split.R (hunk 5 above)
Comment (romainfrancois) [reply R/group_split.R]: wrote an attempt ...

File: R/group_split.R [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+#' @family grouping functions
+#'
+#' [group_split()] works like [base::split()] but
+#' - it uses the grouping structure
+#'   from [group_by()]. In particular, groups may be empty and lead to tibbles with
+#'   0 rows.
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_split.R]: Thanks. I guess we can remove the "In particular ... " sentence then.

**Comment by DavisVaughan** [comment]
I'm really happy with where this is going. I like that it feels more like an extension of the group family than anything completely new.

I'm testing the "dumb" cases, is this the best error message for passing an ungrouped data frame and no `...`? Perhaps checking for the case of empty `...` + ungrouped data frame in `group_split_impl()` would solve it?

``` r
library(dplyr)

iris %>%
  group_nest()
#> Error: `.data` is a corrupt grouped_df, the `"groups"` attribute must be a data frame

iris %>%
  group_keys()
#> # A tibble: 1 x 0

iris %>%
  group_split()
#> Error: `.data` is a corrupt grouped_df, the `"groups"` attribute must be a data frame
```

**Comment by hadley** [comment]
I think `iris %>% group_nest()` should work in the sense of returning a length-1 list.

**Comment by romainfrancois** [comment]
I have added a minimal sketch for `group_map` that does: 

``` r
library(dplyr)
group_map <- dplyr:::group_map

iris %>%
  group_by(Species) %>%
  group_map(~head(.x, 2))
#> # A tibble: 6 x 5
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
#>          <dbl>       <dbl>        <dbl>       <dbl> <fct>     
#> 1          5.1         3.5          1.4         0.2 setosa    
#> 2          4.9         3            1.4         0.2 setosa    
#> 3          7           3.2          4.7         1.4 versicolor
#> 4          6.4         3.2          4.5         1.5 versicolor
#> 5          6.3         3.3          6           2.5 virginica 
#> 6          5.8         2.7          5.1         1.9 virginica

iris %>%
  group_by(Species) %>%
  filter(Species == "setosa") %>%
  group_map(~ mutate(.y, n = nrow(.x)))
#> # A tibble: 3 x 2
#>   Species        n
#>   <fct>      <int>
#> 1 setosa        50
#> 2 versicolor     0
#> 3 virginica      0
```

<sup>Created on 2018-11-16 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by hadley** [comment]
Is this ready for review?

**Comment by romainfrancois** [comment]
Almost. I'll try to address @DavisVaughan comments so that both `group_split(iris)` and `group_nest(iris)` work

**Comment by romainfrancois** [comment]
``` r
library(dplyr)

group_nest(iris)
#> # A tibble: 1 x 1
#>   data              
#>   <list>            
#> 1 <tibble [150 × 5]>
group_split(iris)
#> [[1]]
#> # A tibble: 150 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 140 more rows
```

<sup>Created on 2018-11-19 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by romainfrancois** [comment]
I would say it is reviewable now @hadley

File: R/group_map.R [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#'
+#'   If a __formula__, e.g. `~ head(.x)`, it is converted to a function. In the formula,
+#'   you can use `.` or `.x` to refer to the subset of rows of `.tbl`
+#'   for the given group, and `.y` to refer to the key, a one row tibble that
+#'   identify the group
+#'
+#' @param ... Additional arguments passed on to functions.
+#'
+#' @return The function specified in `.f` is called on each group, and the data frames
+#'         are combined with [bind_rows()]
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_map.R]: In that case perhaps 
- `.x` should only contain the the non grouping variables, i.e. the same thing as we'd find in the list column of a `group_nest()`
- the output of the lambda should be a data frame that does not include the grouping variables
- the grouping variables are added afterwards, along with the `"groups"` attribute that correctly recalculates it's `.rows` columns based on the number of rows of each result

File: R/group_split.R [hunk 8]
```diff
@@ -0,0 +1,78 @@
+#' Split data frame by groups
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_split.R]: In the same file you mean ?

[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_split.R]: yeah that makes a lot of sense. ✅

File: R/group_split.R [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  group_split_impl(group_by(.tbl, ...), environment())
+}
+
+#' @export
+group_split.rowwise_df <- function(.tbl, ...) {
+  if (dots_n(...)) {
+    warn("... is ignored in group_split(<grouped_df>), please use group_by(..., add = TRUE) %>% group_split()")
+  }
+  n <- nrow(.tbl)
+  map(seq_len(n), function(i) structure(.tbl[i, ], class = c("tbl_df", "tbl", "data.frame")))
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_split.R]: done, but each of them is still a rowwise_df

**Comment by romainfrancois** [comment]
closes #3966

File: R/group_keys.R [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+#' @export
+group_keys.data.frame <- function(.tbl, ...){
+  group_keys_impl(group_by(.tbl, ...))
+}
+
+#' @export
+group_keys.grouped_df <- function(.tbl, ...) {
+  if (dots_n(...)) {
+    warn("... is ignored in group_keys(<grouped_df>), please use group_by(..., add = TRUE) %>% group_split()")
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/group_keys.R]: Thanks. Seen this too late to apply the suggested commit.

File: R/grouped-df.r [hunk 11]
```diff
@@ -10,9 +10,6 @@
 #' @param drop deprecated
 #' @export
 grouped_df <- function(data, vars, drop) {
-  if (length(vars) == 0) {
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/grouped-df.r]: The idea is that e.g. `group_by(iris)` produces a grouped_df

File: src/group_indices.cpp [hunk 12]
```diff
@@ -570,9 +571,19 @@ SymbolVector GroupedDataFrame::group_vars(SEXP x) {
 DataFrame grouped_df_impl(DataFrame data, SymbolVector symbols) {
   DataFrame copy(shallow_copy(data));
   set_class(copy, GroupedDataFrame::classes());
-  if (!symbols.size())
-    stop("no variables to group by");
-  GroupedDataFrame::set_groups(copy, build_index_cpp(copy, symbols));
+
+  if (!symbols.size()) {
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply src/group_indices.cpp]: Maybe that can move down to `build_index_cpp`

**Comment by romainfrancois** [comment]
closes #3966

**Comment by romainfrancois** [comment]
current status of `group_map()`: 

``` r
library(dplyr)

mtcars %>% 
  group_by(cyl) %>%
  group_map(~ head(.x, 2L))
#> # A tibble: 6 x 11
#> # Groups:   cyl [3]
#>     cyl   mpg  disp    hp  drat    wt  qsec    vs    am  gear  carb
#> * <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1     4  22.8  108     93  3.85  2.32  18.6     1     1     4     1
#> 2     4  24.4  147.    62  3.69  3.19  20       1     0     4     2
#> 3     6  21    160    110  3.9   2.62  16.5     0     1     4     4
#> 4     6  21    160    110  3.9   2.88  17.0     0     1     4     4
#> 5     8  18.7  360    175  3.15  3.44  17.0     0     0     3     2
#> 6     8  14.3  360    245  3.21  3.57  15.8     0     0     3     4

mtcars %>%
  group_by(cyl) %>%
  group_map(~ tibble(mod = list(lm(mpg ~ disp, data = .x))))
#> # A tibble: 3 x 2
#> # Groups:   cyl [3]
#>     cyl mod     
#> * <dbl> <list>  
#> 1     4 <S3: lm>
#> 2     6 <S3: lm>
#> 3     8 <S3: lm>
```

<sup>Created on 2018-12-03 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by PMassicotte** [comment]
Hi everyone. First, good job on all this. You make my life much simpler :)

Just looking at the examples in the `group_nest()`, is there a difference between these two variations:

```r
iris %>%
  group_by(Species) %>%
  filter(Sepal.Length > mean(Sepal.Length)) %>%
  group_nest()

iris %>%
  group_by(Species) %>%
  filter(Sepal.Length > mean(Sepal.Length)) %>%
  nest()
```

**Comment by romainfrancois** [comment]
There shouldn't be a difference, although I need to fix https://github.com/tidyverse/tidyr/pull/511

**Pull request merged**